charms/jupyter-controller/tests/unit/test_operator.py

```diff
@@ -4,117 +4,172 @@
+        assert harness.charm.model.unit.status == ActiveStatus("")
+
+    def test_prometheus_data_set(self, harness: Harness, mocker):
+        """Test Prometheus data setting."""
+        harness.set_leader(True)
+        harness.set_model_name("test_kubeflow")
+        harness.begin()
+
+        mock_net_get = mocker.patch("ops.testing._TestingModelBackend.network_get")
+
+        bind_address = "1.1.1.1"
+        fake_network = {
+            "bind-addresses": [
+                {
+                    "interface-name": "eth0",
+                    "addresses": [{"hostname": "cassandra-tester-0", "value": bind_address}],
+                }
+            ]
+        }
+        mock_net_get.return_value = fake_network
+
+        rel_id = harness.add_relation("metrics-endpoint", "otherapp")
+        harness.add_relation_unit(rel_id, "otherapp/0")
+        harness.update_relation_data(rel_id, "otherapp", {})
+
+        # basic data
+        assert json.loads(
+            harness.get_relation_data(rel_id, harness.model.app.name)["scrape_jobs"]
+        )[0]["static_configs"][0]["targets"] == ["*:8080"]
+
+        # load alert rules from rules files
+        test_alerts = []
+        with open("src/prometheus_alert_rules/controller.rule") as f:
+            file_alert = yaml.safe_load(f.read())
+            test_alerts.append(file_alert["alert"])
+        with open("src/prometheus_alert_rules/host_resources.rules") as f:
+            file_alert = yaml.safe_load(f.read())
+            # there 2 alert rules in host_resources.rules
+            for rule in file_alert["groups"][0]["rules"]:
+                test_alerts.append(rule["alert"])
+        with open("src/prometheus_alert_rules/model_errors.rule") as f:
+            file_alert = yaml.safe_load(f.read())
+            test_alerts.append(file_alert["alert"])
+        with open("src/prometheus_alert_rules/unit_unavailable.rule") as f:
+            file_alert = yaml.safe_load(f.read())
+            test_alerts.append(file_alert["alert"])
+
+        # alert rules
+        alert_rules = json.loads(
+            harness.get_relation_data(rel_id, harness.model.app.name)["alert_rules"]
+        )
+        assert alert_rules is not None
+        assert alert_rules["groups"] is not None
+
+        # there are 5 alerts
+        rules = []
+        for group in alert_rules["groups"]:
+            for rule in group["rules"]:
+                rules.append(rule)
+
+        # verify number of alerts is the same in relation and in the rules file
+        assert len(rules) == len(test_alerts)
+
+        # verify alerts in relation match alerts in the rules file
+        for rule in rules:
+            assert rule["alert"] in test_alerts
+
+    @patch("charm.JupyterController.k8s_resource_handler")
+    @patch("charm.JupyterController.crd_resource_handler")
+    def test_pebble_layer(
+        self,
+        _: MagicMock,  # k8s_resource_handler
+        __: MagicMock,  # crd_resource_handler
```

Comment:
I see the idea but that is very implicit. You resolve names with comments
```suggestion
        _k8s_resource_handler: MagicMock, 
        _crd_resource_handler: MagicMock,  
```

Reply:
I would have expected using the actual names here would fail our formatting/linting because the `_k8s_resource_handler` etc are unused.  I'm a bit confused why it doesn't fail

Reply:
oh, the leading underscore suppresses the lint issue.  Ok yeah I like this then